### PR TITLE
fix RunTogether construction for e2e tests

### PR DESCRIPTION
**Pull request by matthyx**
#### What type of PR is this?
/kind failing-test

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by matthyx** [comment]
@kannon92 pushing a fix soon...

**Comment by kannon92** [comment]
/hold

Please make sure /test pull-kubernetes-node-swap-fedora is passing before merge.

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [inline test/e2e_node/container_lifecycle_pod_construction.go:132]: This assumes that lhs *always* starts before rhs. Can you point me out why do you think it does?

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [inline test/e2e_node/container_lifecycle_pod_construction.go:135]: The same here - the code assumes that rhs always finished after lhs.

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [inline test/e2e_node/container_lifecycle_pod_construction.go:137]: This should be checked right after lhstart is set. the same for rhsStart check below.

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [inline test/e2e_node/container_lifecycle_pod_construction.go:128]: Can you explain why do we need this loop?

**Review by bart0sh — CHANGES_REQUESTED** [review]

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:132]: because it was intended like that by @tzneal and it makes sense for init containers (which sidecars are)

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:135]: yes, good point - I'll fix that

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: the loop is here because we have some scenarios where the sidecar should be restarting after an early crash and we're catching the run together at a later stage - this is a way to make the test more resilient (look at the failed logs)

Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: the real issue here was that before my PR merged yesterday, the function was looking for "Finishing" instead of "Exiting" which made the check return nil and the test to succeed all the time

Comment (bart0sh) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: Can you point me out to the test case for this scenario?

Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: https://github.com/kubernetes/kubernetes/blob/95fb4224e6130c9c5a57af817492569034d65e23/test/e2e_node/container_lifecycle_test.go#L2003

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [reply test/e2e_node/container_lifecycle_pod_construction.go:132]: So for these init containers:
```
				RestartPolicy: v1.RestartPolicyNever,
				InitContainers: []v1.Container{
					{
						Name:  init1,
						Image: busyboxImage,
						Command: ExecCommand(init1, execCommand{
							Delay:    1,
							ExitCode: 0,
						}),
					},
					{
						Name:  restartableInit1,
						Image: busyboxImage,
						Command: ExecCommand(restartableInit1, execCommand{
							Delay:    600,
							ExitCode: 0,
						}),
						RestartPolicy: &containerRestartPolicyAlways,
					},
					{
						Name:  init2,
						Image: busyboxImage,
						Command: ExecCommand(init2, execCommand{
							Delay:    1,
							ExitCode: 0,
						}),
					},
					{
						Name:  restartableInit2,
						Image: busyboxImage,
						Command: ExecCommand(restartableInit2, execCommand{
							Delay:    600,
							ExitCode: 0,
						}),
						RestartPolicy: &containerRestartPolicyAlways,
					},
					{
						Name:  init3,
						Image: busyboxImage,
						Command: ExecCommand(init3, execCommand{
							Delay:    1,
							ExitCode: 0,
						}),
					},
				},
```
we expect `restartableInit2` to always start before `restartableInit2`. Can you explain why? They have the same Delay, so can start at approximately the same time, no?

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: this is the output before the loop:
```
0) 2024-04-24 13:30:34 +0000 UTC 1333.440000 restartable-init-1 Starting
1) 2024-04-24 13:30:34 +0000 UTC 1333.450000 restartable-init-1 Started
2) 2024-04-24 13:30:34 +0000 UTC 1333.460000 restartable-init-1 Delaying
3) 2024-04-24 13:30:35 +0000 UTC 1334.480000 restartable-init-1 Exiting
4) 2024-04-24 13:30:35 +0000 UTC 1334.620000 init-1 Starting
5) 2024-04-24 13:30:35 +0000 UTC 1334.620000 init-1 Started
6) 2024-04-24 13:30:35 +0000 UTC 1334.630000 init-1 Delaying
7) 2024-04-24 13:30:36 +0000 UTC 1335.630000 restartable-init-1 Starting
8) 2024-04-24 13:30:36 +0000 UTC 1335.640000 init-1 Exiting
9) 2024-04-24 13:30:36 +0000 UTC 1335.650000 restartable-init-1 Started
10) 2024-04-24 13:30:36 +0000 UTC 1335.660000 restartable-init-1 Delaying
11) 2024-04-24 13:30:37 +0000 UTC 1336.610000 init-1 Starting
12) 2024-04-24 13:30:37 +0000 UTC 1336.620000 init-1 Started
13) 2024-04-24 13:30:37 +0000 UTC 1336.620000 init-1 Delaying
14) 2024-04-24 13:30:37 +0000 UTC 1336.660000 restartable-init-1 Exiting
15) 2024-04-24 13:30:38 +0000 UTC 1337.630000 init-1 Exiting
16) 2024-04-24 13:30:54 +0000 UTC 1353.490000 restartable-init-1 Starting
17) 2024-04-24 13:30:54 +0000 UTC 1353.500000 restartable-init-1 Started
18) 2024-04-24 13:30:54 +0000 UTC 1353.520000 restartable-init-1 Delaying
19) 2024-04-24 13:30:54 +0000 UTC 1353.660000 init-1 Started
20) 2024-04-24 13:30:54 +0000 UTC 1353.660000 init-1 Starting
21) 2024-04-24 13:30:54 +0000 UTC 1353.670000 init-1 Delaying
22) 2024-04-24 13:30:55 +0000 UTC 1354.530000 restartable-init-1 Exiting
23) 2024-04-24 13:30:55 +0000 UTC 1354.680000 init-1 Exiting
24) 2024-04-24 13:31:23 +0000 UTC 1382.470000 restartable-init-1 Starting
25) 2024-04-24 13:31:23 +0000 UTC 1382.490000 restartable-init-1 Started
26) 2024-04-24 13:31:23 +0000 UTC 1382.490000 restartable-init-1 Delaying
27) 2024-04-24 13:31:23 +0000 UTC 1382.620000 init-1 Starting
28) 2024-04-24 13:31:23 +0000 UTC 1382.630000 init-1 Started
29) 2024-04-24 13:31:23 +0000 UTC 1382.630000 init-1 Delaying
In [It] at: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:2003 @ 04/24/24 13:31:23.529
```

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:132]: they should not start in parallel, but one after the other once the previous one has exited (non-sidecar) or is ready (restartable=sidecar)

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: ok, thanks. That probably makes sense, but I still don't understand how it worked before your fix.

Comment (matthyx) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: it was working because we never checked the proper ordering https://github.com/kubernetes/kubernetes/pull/124086/files#diff-e363a84d59ad2c9d6634456fa26efa9a53d1c7f32b1f1cfe4871ad9c25964e59L130

Comment (bart0sh) [reply test/e2e_node/container_lifecycle_pod_construction.go:128]: And I also think that looping makes the code less understandable for less complex test cases, when containers don't restart.

File: test/e2e_node/container_lifecycle_pod_construction.go
Comment (bart0sh) [reply test/e2e_node/container_lifecycle_pod_construction.go]: Makes sense. Thanks for the explanations!
However, I'm still wondering how this test case worked in the code in master branch, considering this:
- https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1168
- https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_pod_construction.go#L128

**Comment by bart0sh** [comment]
@matthyx I run `Lifecycle` test cases locally in this PR branch and 5 failed:
```
_output/local/go/bin/e2e_node.test --ginkgo.focus='Containers Lifecycle' --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-swap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --v=7 --kubelet-config-file='test/e2e_node/jenkins/default-kubelet-config.yaml'
...
Summarizing 5 Failures:
  [FAIL] [sig-node] [NodeConformance] Containers Lifecycle when a pod is terminating because its liveness probe fails [It] should execute readiness probe while in preStop, but not liveness [NodeConformance] [sig-node, NodeConformance]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:837
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle when using a restartable init container in a Pod with restartPolicy=Never when a restartable init container starts and exits with exit code 0 continuously [It] should restart a restartable init container after the regular container started [sig-node, NodeFeature:SidecarContainers]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1371
  [FAIL] [sig-node] [NodeConformance] Containers Lifecycle [It] shouldn't restart init containers upon regular container restart [sig-node, NodeConformance]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:533
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle when using a restartable init container in a Pod with restartPolicy=Never when a restartable init container starts and exits with exit code 1 continuously [It] should restart a restartable init container after the regular container started [sig-node, NodeFeature:SidecarContainers]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1443
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle when using a restartable init container in a Pod with restartPolicy=Never when an Init container after restartable init container fails [It] should be running restartable init container and a failed Init container in parallel [sig-node, NodeFeature:SidecarContainers]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1580

Ran 91 of 575 Specs in 1403.254 seconds
FAIL! -- 86 Passed | 5 Failed | 0 Pending | 484 Skipped
```

**Comment by matthyx** [comment]
Ok maybe I should just revert yesterday's PR and take more time to deflake.
Sorry I'm not available atm, can you open a revert please?

**Comment by bart0sh** [comment]
Sure, can you point me out to the PR?

**Comment by bart0sh** [comment]
I guess it's this one: https://github.com/kubernetes/kubernetes/pull/124086

**Comment by bart0sh** [comment]
I'll revert it locally and run tests again.

**Comment by bart0sh** [comment]
With reverted https://github.com/kubernetes/kubernetes/pull/124086 it still fails:
```
Summarizing 2 Failures:
  [FAIL] [sig-node] [NodeConformance] Containers Lifecycle when a pod is terminating because its liveness probe fails [It] should execute readiness probe while in preStop, but not liveness [NodeConformance] [sig-node, NodeConformance]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:837
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle when using a restartable init container in a Pod with restartPolicy=Always when a restartable init container runs continuously [It] should not restart a restartable init container [sig-node, NodeFeature:SidecarContainers]
  /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:2062

Ran 75 of 575 Specs in 1301.554 seconds
FAIL! -- 73 Passed | 2 Failed | 0 Pending | 500 Skipped
```

**Comment by bart0sh** [comment]
Yes, it fails on master with reverted https://github.com/kubernetes/kubernetes/pull/124086, but it doesn't fail `using a Pod with restartPolicy=Never, three init container and two restartable init containers should run both restartable init containers and third init container together` test case, so I've submitted PR as you've suggested: https://github.com/kubernetes/kubernetes/pull/124512

**Comment by matthyx** [comment]
> Yes, it fails on master with reverted #124086, but it doesn't fail `using a Pod with restartPolicy=Never, three init container and two restartable init containers should run both restartable init containers and third init container together` test case, so I've submitted PR as you've suggested: #124512

Thanks you!

**Comment by matthyx** [comment]
@SergeyKanzhelev can you approve, I will rework my PR soon - sorry for the failures

**Comment by SergeyKanzhelev** [comment]
/lgtm
/approve

I also approved the revert of tests. This may end up conflicting with it.

/hold

feel free to unhold

**Comment by matthyx** [comment]
> @SergeyKanzhelev can you approve, I will rework my PR soon - sorry for the failures

oops, I meant the revert... I need to rework it making sure it doesn't flake.
Please keep hold on it for the moment.

**Comment by bart0sh** [comment]
@matthyx I'm trying to fix remaining container lifecycle test failures. Here is one [failing test case](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1211) that fails in a strange way:

```
[sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle when using a restartable init container in a Pod with restartPolicy=Never when a restartable init container starts and exits with exit code 0 continuously [It] should restart a restartable init container after the regular container started [sig-node, NodeFeature:SidecarContainers]
kubernetes/test/e2e_node/container_lifecycle_test.go:1211

  [FAILED] couldn't find that restartable-init-1 started after regular-1, got
  0) 2024-04-25 19:27:15 +0300 EEST 28443028.120000 restartable-init-1 Starting
  1) 2024-04-25 19:27:15 +0300 EEST 28443028.130000 restartable-init-1 Started
  2) 2024-04-25 19:27:15 +0300 EEST 28443028.130000 restartable-init-1 Delaying
  3) 2024-04-25 19:27:16 +0300 EEST 28443028.360000 init-1 Starting
  4) 2024-04-25 19:27:16 +0300 EEST 28443028.370000 init-1 Started
  5) 2024-04-25 19:27:16 +0300 EEST 28443028.380000 init-1 Delaying
  6) 2024-04-25 19:27:16 +0300 EEST 28443029.150000 restartable-init-1 Exiting
  7) 2024-04-25 19:27:18 +0300 EEST 28443030.400000 restartable-init-1 Starting
  8) 2024-04-25 19:27:18 +0300 EEST 28443030.410000 restartable-init-1 Started
  9) 2024-04-25 19:27:18 +0300 EEST 28443030.420000 restartable-init-1 Delaying
  10) 2024-04-25 19:27:19 +0300 EEST 28443031.430000 restartable-init-1 Exiting
  11) 2024-04-25 19:27:21 +0300 EEST 28443033.390000 init-1 Exiting
  12) 2024-04-25 19:27:22 +0300 EEST 28443034.490000 regular-1 Starting
  13) 2024-04-25 19:27:22 +0300 EEST 28443034.500000 regular-1 Started
  14) 2024-04-25 19:27:22 +0300 EEST 28443034.510000 regular-1 Delaying
  15) 2024-04-25 19:27:37 +0300 EEST 28443049.390000 regular-1 Exiting
  
  In [It] at: /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1212 @ 04/25/24 19:27:41.217
```

Do you have any idea why delaying `regular-1` container took 15 seconds(lines 15-16 above) instead of [60](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1183)?

[Similar test case](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1283) `a restartable init container starts and exits with exit code 1 continuously` fails the same way.

**Comment by matthyx** [comment]
> @matthyx I'm trying to fix remaining container lifecycle test failures. Here is one [failing test case](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1211) that fails in a strange way:
> 
> ```
> [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle when using a restartable init container in a Pod with restartPolicy=Never when a restartable init container starts and exits with exit code 0 continuously [It] should restart a restartable init container after the regular container started [sig-node, NodeFeature:SidecarContainers]
> kubernetes/test/e2e_node/container_lifecycle_test.go:1211
> 
>   [FAILED] couldn't find that restartable-init-1 started after regular-1, got
>   0) 2024-04-25 19:27:15 +0300 EEST 28443028.120000 restartable-init-1 Starting
>   1) 2024-04-25 19:27:15 +0300 EEST 28443028.130000 restartable-init-1 Started
>   2) 2024-04-25 19:27:15 +0300 EEST 28443028.130000 restartable-init-1 Delaying
>   3) 2024-04-25 19:27:16 +0300 EEST 28443028.360000 init-1 Starting
>   4) 2024-04-25 19:27:16 +0300 EEST 28443028.370000 init-1 Started
>   5) 2024-04-25 19:27:16 +0300 EEST 28443028.380000 init-1 Delaying
>   6) 2024-04-25 19:27:16 +0300 EEST 28443029.150000 restartable-init-1 Exiting
>   7) 2024-04-25 19:27:18 +0300 EEST 28443030.400000 restartable-init-1 Starting
>   8) 2024-04-25 19:27:18 +0300 EEST 28443030.410000 restartable-init-1 Started
>   9) 2024-04-25 19:27:18 +0300 EEST 28443030.420000 restartable-init-1 Delaying
>   10) 2024-04-25 19:27:19 +0300 EEST 28443031.430000 restartable-init-1 Exiting
>   11) 2024-04-25 19:27:21 +0300 EEST 28443033.390000 init-1 Exiting
>   12) 2024-04-25 19:27:22 +0300 EEST 28443034.490000 regular-1 Starting
>   13) 2024-04-25 19:27:22 +0300 EEST 28443034.500000 regular-1 Started
>   14) 2024-04-25 19:27:22 +0300 EEST 28443034.510000 regular-1 Delaying
>   15) 2024-04-25 19:27:37 +0300 EEST 28443049.390000 regular-1 Exiting
>   
>   In [It] at: /home/ed/go/src/k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:1212 @ 04/25/24 19:27:41.217
> ```
> 
> Do you have any idea why delaying `regular-1` container took 15 seconds(lines 15-16 above) instead of [60](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1183)?
> 
> [Similar test case](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1283) `a restartable init container starts and exits with exit code 1 continuously` fails the same way.

@bart0sh this is very strange... how are you running your tests? On my Linux computer I get this log (with a simulated error):
```
  [FAILED] fake error, got
  0) 2024-04-26 07:39:35 +0200 CEST 1814362.120000 restartable-init-1 Starting
  1) 2024-04-26 07:39:35 +0200 CEST 1814362.130000 restartable-init-1 Started
  2) 2024-04-26 07:39:35 +0200 CEST 1814362.130000 restartable-init-1 Delaying
  3) 2024-04-26 07:39:36 +0200 CEST 1814363.080000 init-1 Starting
  4) 2024-04-26 07:39:36 +0200 CEST 1814363.090000 init-1 Delaying
  5) 2024-04-26 07:39:36 +0200 CEST 1814363.090000 init-1 Started
  6) 2024-04-26 07:39:36 +0200 CEST 1814363.130000 restartable-init-1 Exiting
  7) 2024-04-26 07:39:37 +0200 CEST 1814364.080000 restartable-init-1 Starting
  8) 2024-04-26 07:39:37 +0200 CEST 1814364.080000 restartable-init-1 Started
  9) 2024-04-26 07:39:37 +0200 CEST 1814364.090000 restartable-init-1 Delaying
  10) 2024-04-26 07:39:38 +0200 CEST 1814365.100000 restartable-init-1 Exiting
  11) 2024-04-26 07:39:41 +0200 CEST 1814368.110000 init-1 Exiting
  12) 2024-04-26 07:39:42 +0200 CEST 1814369.210000 regular-1 Starting
  13) 2024-04-26 07:39:42 +0200 CEST 1814369.220000 regular-1 Started
  14) 2024-04-26 07:39:42 +0200 CEST 1814369.230000 regular-1 Delaying
  15) 2024-04-26 07:39:55 +0200 CEST 1814383.010000 restartable-init-1 Starting
  16) 2024-04-26 07:39:56 +0200 CEST 1814383.020000 restartable-init-1 Started
  17) 2024-04-26 07:39:56 +0200 CEST 1814383.020000 restartable-init-1 Delaying
  18) 2024-04-26 07:39:57 +0200 CEST 1814384.030000 restartable-init-1 Exiting
  19) 2024-04-26 07:40:23 +0200 CEST 1814410.040000 restartable-init-1 Starting
  20) 2024-04-26 07:40:23 +0200 CEST 1814410.050000 restartable-init-1 Delaying
  21) 2024-04-26 07:40:23 +0200 CEST 1814410.050000 restartable-init-1 Started
  22) 2024-04-26 07:40:24 +0200 CEST 1814411.060000 restartable-init-1 Exiting
  23) 2024-04-26 07:40:42 +0200 CEST 1814429.240000 regular-1 Exiting
```

**Comment by bart0sh** [comment]
It's very strange indeed. I'm running the tests this way:
```
$ sudo _output/local/go/bin/e2e_node.test --ginkgo.focus='Containers Lifecycle' --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-swap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --v=3 --kubelet-config-file='test/e2e_node/jenkins/default-kubelet-config.yaml'
```

**Comment by matthyx** [comment]
@bart0sh I just ran 30 times the test on my computer (watching TV) and every single time I had exactly 1 minute between `Delaying` and `Exiting`.

However I do see different orders in the lines that are printed with the same timestamp:
- `Starting`- `Started` - `Delaying`
- `Delaying` - `Started` - `Starting`
- etc...

Maybe something to improve in the framework?

**Comment by bart0sh** [comment]
> Starting- Started - Delaying
> Delaying - Started - Starting
> etc...

This looks like a correct output of [this code](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_pod_construction.go#L79-L88) when StartDelay is 0.

**Comment by matthyx** [comment]
> > Starting- Started - Delaying
> > Delaying - Started - Starting
> > etc...
> 
> This looks like a correct output of [this code](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_pod_construction.go#L79-L88) when StartDelay is 0.

Yeah, it's just that sometimes the order is different which might cause issues if we start parsing for specific behaviors...

**Comment by bart0sh** [comment]
@matthyx 
> Yeah, it's just that sometimes the order is different which might cause issues if we start parsing for specific behaviors...

This is caused by [sorting parsed output](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_pod_construction.go#L356), I believe. I saw this in my environment as well:

[FAILED] couldn't find that restartable-init-1 started after regular-1, got
  0) 2024-04-28 19:01:46 +0300 EEST 28700699.200000 restartable-init-1 Starting
  1) 2024-04-28 19:01:47 +0300 EEST 28700699.210000 restartable-init-1 Started
  2) 2024-04-28 19:01:47 +0300 EEST 28700699.210000 restartable-init-1 Delaying
  3) 2024-04-28 19:01:47 +0300 EEST 28700700.200000 init-1 Starting
  4) 2024-04-28 19:01:48 +0300 EEST 28700700.210000 init-1 Delaying
  5) 2024-04-28 19:01:48 +0300 EEST 28700700.210000 init-1 Started
  6) 2024-04-28 19:01:48 +0300 EEST 28700700.220000 restartable-init-1 Exiting
  7) 2024-04-28 19:01:48 +0300 EEST 28700701.120000 **restartable-init-1 Starting**
  8) 2024-04-28 19:01:48 +0300 EEST 28700701.130000 **restartable-init-1 Delaying**
  9) 2024-04-28 19:01:48 +0300 EEST 28700701.130000 **restartable-init-1 Started**
  10) 2024-04-28 19:01:49 +0300 EEST 28700702.140000 **restartable-init-1 Exiting**
  11) 2024-04-28 19:01:53 +0300 EEST 28700705.220000 init-1 Exiting
  12) 2024-04-28 19:01:53 +0300 EEST 28700706.180000 regular-1 Starting
  13) 2024-04-28 19:01:53 +0300 EEST 28700706.190000 regular-1 Started
  14) 2024-04-28 19:01:54 +0300 EEST 28700706.200000 regular-1 Delaying
  15) 2024-04-28 19:02:07 +0300 EEST 28700720.180000 regular-1 Exiting

Note, that `timeSinceBoot` is the same for the records 8 and 9.

**Comment by gjkim42** [comment]
> @matthyx
> 
> > Yeah, it's just that sometimes the order is different which might cause issues if we start parsing for specific behaviors...
> 
> This is caused by [sorting parsed output](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_pod_construction.go#L356), I believe. I saw this in my environment as well:
> 
> [FAILED] couldn't find that restartable-init-1 started after regular-1, got 0) 2024-04-28 19:01:46 +0300 EEST 28700699.200000 restartable-init-1 Starting
> 
> 1. 2024-04-28 19:01:47 +0300 EEST 28700699.210000 restartable-init-1 Started
> 2. 2024-04-28 19:01:47 +0300 EEST 28700699.210000 restartable-init-1 Delaying
> 3. 2024-04-28 19:01:47 +0300 EEST 28700700.200000 init-1 Starting
> 4. 2024-04-28 19:01:48 +0300 EEST 28700700.210000 init-1 Delaying
> 5. 2024-04-28 19:01:48 +0300 EEST 28700700.210000 init-1 Started
> 6. 2024-04-28 19:01:48 +0300 EEST 28700700.220000 restartable-init-1 Exiting
> 7. 2024-04-28 19:01:48 +0300 EEST 28700701.120000 **restartable-init-1 Starting**
> 8. 2024-04-28 19:01:48 +0300 EEST 28700701.130000 **restartable-init-1 Delaying**
> 9. 2024-04-28 19:01:48 +0300 EEST 28700701.130000 **restartable-init-1 Started**
> 10. 2024-04-28 19:01:49 +0300 EEST 28700702.140000 **restartable-init-1 Exiting**
> 11. 2024-04-28 19:01:53 +0300 EEST 28700705.220000 init-1 Exiting
> 12. 2024-04-28 19:01:53 +0300 EEST 28700706.180000 regular-1 Starting
> 13. 2024-04-28 19:01:53 +0300 EEST 28700706.190000 regular-1 Started
> 14. 2024-04-28 19:01:54 +0300 EEST 28700706.200000 regular-1 Delaying
> 15. 2024-04-28 19:02:07 +0300 EEST 28700720.180000 regular-1 Exiting
> 
> Note, that `timeSinceBoot` is the same for the records 8 and 9.

It may work better with stable sort to keep the original ordering if the timestamp are same.
https://pkg.go.dev/sort#Stable

**Comment by matthyx** [comment]
taking a different approach

**Pull request closed without merging**
